tests/integration/terraform_setup.py

```diff
@@ -0,0 +1,192 @@
+    """Ensure Terraform is installed and matches the expected version.
+
+    Args:
+        expected_version: The expected version of Terraform.
+    """
+    installed_version = run_command(
+        "snap list terraform | awk '/^terraform/ {print $4}'", capture_output=True
+    )
+    if not installed_version:
+        print(f"Terraform is not installed. Installing version {expected_version}...")
+        run_command(f"sudo snap install terraform --channel={expected_version} --classic")
+    elif installed_version != expected_version:
+        print(
+            f"Error: Installed Terraform version ({installed_version})"
+            f"does not match the expected version ({expected_version})."
+        )
+        sys.exit(1)
+    else:
+        print(
+            f"Terraform is already installed and matches the expected version: {installed_version}."
+        )
+
+
+def setup_juju_provider_authentication() -> None:
+    """Set up Juju provider authentication.
+
+    See https://registry.terraform.io/providers/juju/juju/latest/docs#authentication
+    """
+    controller = run_command("juju whoami | yq .Controller", capture_output=True) or ""
+    os.environ["CONTROLLER"] = controller or ""
+    os.environ["JUJU_CONTROLLER_ADDRESSES"] = (
+        run_command(
+            f"juju show-controller | yq .{controller}.details.api-endpoints"
+            "| yq -r '. | join(\",\")'",
+            capture_output=True,
+        )
+        or ""
+    )
+    os.environ["JUJU_USERNAME"] = (
+        run_command(
+            f"cat ~/.local/share/juju/accounts.yaml"
+            f"| yq .controllers.{controller}.user | tr -d '\"'",
+            capture_output=True,
+        )
+        or ""
+    )
+    os.environ["JUJU_PASSWORD"] = (
+        run_command(
+            f"yq .controllers.{controller}.password < ~/.local/share/juju/accounts.yaml",
+            capture_output=True,
+        )
+        or ""
+    )
+    os.environ["JUJU_CA_CERT"] = (
+        run_command(
+            f"juju show-controller {controller.strip()}"
+            f"| yq '.{controller}.details.\"ca-cert\"' | tr -d '\"' | sed 's/\\\\n/\\n/g'",
+            capture_output=True,
+        )
+        or ""
+    )
+
+
+def ensure_model_exists(model_name: str, lxd_profile_path: Union[Path, str]) -> None:
+    """Ensure the specified Juju model exists.
+
+    If not, create it and configure LXD profile if required.
+
+    Args:
+        model_name: The name of the Juju model.
+        lxd_profile_path: The path to the LXD profile file.
+    """
+    # Grep returns 1 if no match is found, so we use `fail_on_error=False` to suppress the error
+    if run_command(
+        f"juju models | grep -q {model_name}", capture_output=True, fail_on_error=False
+    ):
+        print(f"Juju model '{model_name}' already exists.")
+    else:
+        print(f"Juju model '{model_name}' does not exist. Creating it...")
+        run_command(f"juju add-model {model_name} localhost")
```

Comment:
Maybe reference the controller here rather than the cloud `localhost`

```suggestion
        run_command(f"juju add-model -c {controller} {model_name}")
```